docs: finish v4 migration doc (#1311)
interactions-py/interactions.py: docs/src/Guides/98 Migration from 4.X.md

## Events

Similarly to Slash Commands, events have also been reworked in v5. Instead of `@bot.event` and `@extension_listener`, the way to listen to events is now `@listen`. There are multiple ways to subscribe to events, whether it is using the function name or the argument of the `@listen` decorator. You can find more information on handling events using v5 [on its own guide page](../10 Events).

## Other Types of Interactions (Context Menus, Components, Modals)

These should be a lot more familiar to you - many interactions in v5 that aren't slash commands are similar to v4, minus name changes (largely to the decorators and classes you use). They should still *function* similarly though, but it's never a bad idea to consult the various guides that are on the sidebar to gain a better picture of how they work.

[If you're using context menus](../04 Context Menus) (IE `@bot.user_command` or `@bot.message_command`), note that the decorator you use is slightly different from either of the two older ones. You now use `@context_menu`, and specify the type of context menu through `context_type` - otherwise, it's mostly the same.

> There also is no "one decorator for every type of command" - there is no equivalent to `bot.command`, and you will need to use the more narrowly-scoped alternatives.

Reviewer: ```suggestion
There also is no "one decorator for every type of command" - there is no equivalent to `bot.command`, and you will need to use the specialized decorators instead. 

For example
```python
@slash_command(...)  # for slash commands
@subcommand(...)  # for slash subcommands
@context_menu(...)  # for context menus
@component_callback(...)  # for component callbacks
```

Autodeferring is also pretty similar, although there's more control, with options to allow for global autodefers and extension-wide ones.

## Events

Similarly to Slash Commands, events have also been reworked in v5. Instead of `@bot.event` and `@extension_listener`, the way to listen to events is now `@listen`. There are multiple ways to subscribe to events, whether it is using the function name or the argument of the `@listen` decorator. You can find more information on handling events using v5 [on its own guide page](../10 Events).

## Other Types of Interactions (Context Menus, Components, Modals)

These should be a lot more familiar to you - many interactions in v5 that aren't slash commands are similar to v4, minus name changes (largely to the decorators and classes you use). They should still *function* similarly though, but it's never a bad idea to consult the various guides that are on the sidebar to gain a better picture of how they work.

> [If you're using context menus](../04 Context Menus) (IE `@bot.user_command` or `@bot.message_command`), note that the decorator you use is slightly different from either of the two older ones. You now use `@context_menu`, and specify the type of context menu through `context_type` - otherwise, it's mostly the same.

Reviewer 2: We added equivalents `message_context_menu` and `user_context_menu` decorators. Can you add info there?

Author: Done, maybe I should add in the Context Menu guide the mention of these two decorators? (I just learned they exist lol)

Reviewer 2: Not in this pr I think, but better ask Polls about it

Reviewer: @Damego @mAxYoLo01 go for it

There also is no "one decorator for every type of command" - there is no equivalent to `bot.command`, and you will need to use the specialized decorators instead.

For example:
```python
@slash_command(...)  # for slash commands
@subcommand(...)  # for slash subcommands
@context_menu(...)  # for context menus

> @component_callback(...)  # for component callbacks

Reviewer 2: ```suggestion
@component_callback(...)  # for component callbacks
@modal_callback(...)  # for modal callbacks
```

## Extensions (cogs)

Extensions have not been changed too much. `await teardown(...)` is now just `drop(...)` (note how drop is *not* async), and you use `bot.load_extension`/`bot.unload_extension` instead of `bot.load`/`bot.unload`.

There is one major difference though that isn't fully related to extensions themselves: *you use the same decorator for both commands/events in your main file and commands/events in extensions in v5.* Basically, instead of having `bot.command` and `interactions.extension_command`, you *just* have `interactions.slash_command` (and so on for context menus, events, etc.), which functions seemlessly in both contexts.

Also, you no longer require a `setup` function. They can still be used, but if you have no need for them other than just loading the extension, you can get rid of them if you want.

## Cache and interactions.get

Instead of the `await interactions.get` function in v4, v5 introduces the `await bot.fetch_X` and `bot.get_X` functions, where `X` will be the type of object that you would like to retrieve (user, guild, role...). You might ask, what is the difference between fetch and get?

> The answer is simple, `get` will only look for an object that has been cached, and therefore is a synchronous function that can return None if this object has never been cached before.

Reviewer: ```suggestion
The answer is simple, `get` will look for an object that has been cached, and therefore is a synchronous function that can return None if this object has never been cached before.
```

> On the other hand, `fetch` is an asynchronous function that will request the Discord API to find that object if it has not been cached before. This ensures that you will always retrieve the most up to date version of that object, provided that the IDs you inputted are valid.

Reviewer: ```suggestion
On the other hand, `fetch` is an asynchronous function that will request the Discord API to find that object if it has not been cached before. This will *fetch* the latest version of the object from Discord, provided that the IDs you inputted are valid.
```
